Optimize waitingforkey event reporting by acknowledging pipeline state

calvaris opened the pull request.

Commit:
MediaPlayer should have mediaPlayerWaitingForKeyChanged() / bool waitingForKey() accessor
https://bugs.webkit.org/show_bug.cgi?id=189951

Reviewed by Eric Carlson.

Source/WebCore:

In order to implement the "Resume Playback" section of EME, part 4, we need to be able
to query whether the MediaPlayer is still waiting for a key after attemptToDecrypt()
has been called. Currently this involves no behavioral changes, as all modern EME ports
will still just notify the media element that they no longer need keys after one has
been added, but future ports may be able to wait for multiple keys before reporting
that it is no longer waiting for keys.

* html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerWaitingForKeyChanged):
(WebCore::HTMLMediaElement::attemptToResumePlaybackIfNecessary):
(WebCore::HTMLMediaElement::mediaPlayerWaitingForKey): Deleted.
* html/HTMLMediaElement.h:
* platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::waitingForKeyChanged):
(WebCore::MediaPlayer::waitingForKey const):
(WebCore::MediaPlayer::waitingForKey): Deleted.
* platform/graphics/MediaPlayer.h:
(WebCore::MediaPlayerClient::mediaPlayerWaitingForKeyChanged):
(WebCore::MediaPlayerClient::mediaPlayerWaitingForKey): Deleted.
* platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::waitingForKey const):
* platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::shouldWaitForLoadingOfResource):
(WebCore::MediaPlayerPrivateAVFoundationObjC::attemptToDecryptWithInstance):
* platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::attemptToDecryptWithInstance):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::waitingForKey const):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::waitingForKeyChanged):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::initializationDataEncountered):
* platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::didProvideContentKeyRequestInitializationDataForTrackID):
(WebCore::SourceBufferPrivateAVFObjC::attemptToDecrypt):
* platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
* platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp:
(WebCore::MediaPlayerPrivateGStreamerBase::reportWaitingForKey):
(WebCore::MediaPlayerPrivateGStreamerBase::setWaitingForKey):
(WebCore::MediaPlayerPrivateGStreamerBase::waitingForKey const):
* platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.h:
* platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp:
(webkitMediaCommonEncryptionDecryptSinkEventHandler):

Source/WTF:

Templated functions should take r-value references, as they have perfect type deduction for
all parameter types; references, l-value references, and r-value references in template function
parameters have special type deduction semantics.
See: <https://en.cppreference.com/w/cpp/language/reference#Forwarding_references>

Previously, const reference parameters would be copied when passed into anyOf(), and containers
of Ref<> would generate compile errors when passed into anyOf, as they cannot be copied. Now,
with r-value reference types in template parameters, a const reference is mapped to a const reference,
a non-const reference is mapped to a non-const reference, and a r-value reference is mapped to
an r-value reference.

* wtf/Algorithms.h:
(WTF::forEach):
(WTF::anyOf):
(WTF::allOf):

git-svn-id: http://svn.webkit.org/repository/webkit/trunk@236572 268f45cc-cd09-0410-ab3c-d52691b4dbfc

calvaris committed:
[EME] Inline reportWaitingForKey

calvaris committed:
[EME] Query decryptors to see if they are waiting at waitingForKey

If they are still waiting, we don't fire the change to false to the
player and element so should not be attempted.

calvaris committed:
[EME] Decryptors handle waitingforkey in the decryption thread

calvaris committed:
[EME] Run waitingforkey query only in >= PAUSED

calvaris committed:
[EME] Query decryptors manually if the pipeline is READY

charlie-ht commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp:
```diff
@@ -1378,10 +1378,47 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionStructure(GUniquePtr<Gst
+            do {
+                iteratorResult = gst_iterator_fold(iterator.get(), [](const GValue *item, GValue *, gpointer) -> gboolean {
+                    GstElement* element = GST_ELEMENT(g_value_get_object(item));
+                    return !WEBKIT_IS_MEDIA_CENC_DECRYPT(element) || !webKitMediaCommonEncryptionDecryptIsWaitingForKey(WEBKIT_MEDIA_CENC_DECRYPT(element));
```
This would read more naturally as return IS_CENC_DECRYPT(element) && isWaitingForKey(element)

calvaris replied:
Yes, but we need to keep the null namespacing somehow.

charlie-ht replied:
I don't understand, I'm applying de morgan here, it would be nice to applying a negation across the conditional and switch it to &&, either way, it's up to you, just a nit :)

calvaris replied:
Me no entiende what you say either :upside_down_face: .

We are doing something similar to what we do with query, we are coalescing the return value of the query being handled and the return value of the query itself because for us it is ok if we find just 1 decryptor that is still blocked to answer `true` so we if just answer `true`, it means that we can stop the query and it means we found a waiting decryptor.

For the case of the iterator, the iterator fold continues when we return `TRUE` in and we want to stop only if we find a decryptor and that decryptor is waiting, so we applying De Morgan we continue if we are not dealing with a decryptor or that decryptor is not waiting.

calvaris replied:
I think I got it now, so you prefer to negate the whole thing and have an and? I don't :smile:

charlie-ht commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp:
```diff
@@ -1378,10 +1378,47 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionStructure(GUniquePtr<Gst
+bool MediaPlayerPrivateGStreamerBase::waitingForKey() const
+{
+    bool result { false };
+    if (m_pipeline) {
```
This looks like quite a slow process, given that the waiting for key status and change in a very short time in practice, my gut feel is by the time this query has been answered, the duration of us actually waiting for a key will be missed. Really need test cases for this.

calvaris replied:
It isn't. I run several test runs and gathered to each test run how long all calls to this function take (the sum of the time it takes calling this function) and the longest case is 8,6ms so a quite acceptable result IMHO for the proper compliance with the spec.

calvaris replied:
And this times are measured on the Rpi2

charlie-ht replied:
That's great :)

charlie-ht reviewed (commented):
Can this be tried upstream with the recent bugzillas I have posted? There are some failing test expectations for waitingforkey up there at the moment, and it seems like the main part of this patch could go upstream and run against the waitingforkey tests there before coming downstream. In general I'm nervous about merging this code without any test coverage, since it looks very sensitive to timing issues.

calvaris commented:
I am running tests upstream right now as well

calvaris commented:
Showtime upstream https://bugs.webkit.org/show_bug.cgi?id=191459

Commit:
[GStreamer][EME] waitingforkey event should consider decryptors' waiting status
https://bugs.webkit.org/show_bug.cgi?id=191459

Reviewed by Carlos Garcia Campos.

The new cross platform architecture to report waitingforkey and
recover from it requires a more accurate knowledge of what is
going on with the decryptors because events are reported only once
(per key exchange run) and crossplatform only continues if we are
actually ready to continue, meaning that no decryptors are
waiting.

* platform/graphics/gstreamer/GUniquePtrGStreamer.h: Added
GstIterator deleter.
* platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp:
(WebCore::MediaPlayerPrivateGStreamerBase::setWaitingForKey): Bail
out if we are requested to not wait anymore but there are still
waiting decryptors.
(WebCore::MediaPlayerPrivateGStreamerBase::waitingForKey const):
Query the pipeline, just a query after pipeline is built and
manual inspection during build. The query is optimal but sometimes
we can get this request when the pipeline is under construction so
queries do not arrive at the decryptors and we have to deliver it
by ourselves.
(WebCore::MediaPlayerPrivateGStreamerBase::reportWaitingForKey:
* platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.h:
(WebCore::MediaPlayerPrivateGStreamerBase::reportWaitingForKey:
Deleted because it is now inlined.
* platform/graphics/gstreamer/eme/WebKitClearKeyDecryptorGStreamer.cpp:
(webKitMediaClearKeyDecryptorDecrypt): Fixed small compiler warning.
* platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp:
(webkit_media_common_encryption_decrypt_class_init): Override
query method.
(webkitMediaCommonEncryptionDecryptTransformInPlace): When the
decryptor is going to block to wait, report before. When the
decryptor receives the key, report it got it.
(webkitMediaCommonEncryptionDecryptSinkEventHandler): Do not
handle waitingforkey here.
(webkitMediaCommonEncryptionDecryptorQueryHandler): Report if the
decryptor is waiting.

git-svn-id: http://svn.webkit.org/repository/webkit/trunk@238083 268f45cc-cd09-0410-ab3c-d52691b4dbfc
